### DISCUSSION TITLE
Scheduled/Automated Print Job Start

### DISCUSSION THREAD

**Discussion by Infraviored** [greghesp/ha-bambulab#628] [Ideas]
## Current Situation
Looking at the integration code, I can see we have pause/resume/stop functionality via MQTT commands (in `commands.py`), but starting a print job isn't implemented yet.

## Use Case
I want to automate print starts based on Home Assistant triggers (e.g., geofencing when leaving home) to avoid printer noise during the day.

The workflow I'm thinking about:
1. Upload file to printer's SD card first (manually or via automation)
2. Start the print job from SD card later through HA automation

Since the Bambu mobile app can already start prints from SD card, this should be possible to implement.

## Technical Investigation
Looking at `commands.py`, I found:
- Communication is via MQTT on port 8883
- Command structure like:
  ```python
  PAUSE = {"print": {"sequence_id": "0", "command": "pause"}}
  RESUME = {"print": {"sequence_id": "0", "command": "resume"}}
  ```
- No existing commands for listing/starting prints from SD card

## Questions
1. Has anyone captured the MQTT traffic between app and printer to see:
   - Commands for accessing the "Models" page (SD card listing)
   - Commands for starting a print
2. Are there any security considerations for implementing file access?
3. Is there existing documentation for these commands?

I'm a Python dev willing to help implement this. Would love to collaborate with others who might have already looked into these MQTT commands or are interested in figuring this out.

**Comment by Infraviored** [comment]
# Update: Progress on Print Start Implementation

I've been monitoring MQTT traffic between my printer and the Bambu app. Here's what I found:

When starting a print via the app, this command appears:
```json
{
  "print": {
    "sequence_id": "2032",
    "command": "project_file",
    "param": "Metadata/plate_1.gcode",
    "subtask_name": "Wirespool_A.3mf",
    "plate_idx": 0,
    "url": "file:///sdcard/cache/Wirespool_A.3mf",
    "timelapse": false,
    "bed_leveling": true,
    "flow_cali": false,
    "vibration_cali": false,
    "layer_inspect": false,
    "ams_mapping": [0, -1],
    "use_ams": true
  }
}
```

However, I'm still missing how to:
1. List available files from the SD card (the command used when opening the Models page)
2. See the complete startup sequence (any commands before/after project_file)

I'm monitoring MQTT on port 8883 with topics:
- `device/{serial}/report` (from printer)
- `device/{serial}/request` (to printer)

Anyone familiar with these commands or can point me to documentation? I'll keep investigating and share what I find.

**Comment by AdrianGarside** [comment]
It is all possible to do, but also quite a bit of work so I haven't got any plans to start work on this any time soon.

**Reply by Infraviored** [reply]
Don't you want to hint me in some direction? I am willing implement this and merge later.

**Reply by AdrianGarside** [reply]
I have notes squirreled away but not sure where now. You have to upload the files to the printer via ftps - I wrote the minimal proof of concept of that in python quite a while ago - it's in a discussion or maybe issue somewhere here. Similarly ftps is how you enumerate the files already on the printer. And you've already found the mqtt command to trigger a print start. Between those it should be possible end to end.

**Reply by RHolmesLboro** [reply]
Which was the MQTT command to trigger a print? I have some code to send any given file over FTPS but can't seem to get the printer to initiate the print, and what arguments are available in the MQTT command (enable AMS, etc)?

**Reply by jneilliii** [reply]
https://github.com/Doridian/OpenBambuAPI/blob/main/mqtt.md#printproject_file

https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/d0fd4a5434043670c3fc02a2ce7020fc885be1d5/octoprint_bambu_printer/printer/states/idle_state.py#L22

**Reply by Infraviored** [reply]
> Which was the MQTT command to trigger a print? I have some code to send any given file over FTPS but can't seem to get the printer to initiate the print, and what arguments are available in the MQTT command (enable AMS, etc)?

Could you share the code you are speaking about?
Are you also trying to implement the feature i am talking about?

I do not really care about uploading files to the printer, as i will typically do this with the slicer in advance.
But Pulling the list of 3mf files together with their description (name, thumbnail) to represent them in Homeassistant, and to start a job using the description is relevant for me.

**Reply by RHolmesLboro** [reply]
Give me 3 hours and remind me, I'll send the code I used to poll the files with python. I can pull down the file names and paths then open them to get the rest of the information you want.

**Reply by Infraviored** [reply]
```
flo@NUC:~/repos/OctoPrint-BambuPrinter$ /bin/python3 /home/flo/repos/OctoPrint-BambuPrinter/octoprint_bambu_printer/printer/file_system/list_3mf_files.py

Connecting to printer at 192.168.178.70...

Scanning for 3MF files...
------------------------------------------------------------
File: /Untitled.gcode.3mf
Name: Untitled.gcode.3mf
DOS Name: untitl~1.3mf
Size: 1,103,439 bytes
Date: 2024-10-09 23:49:42+00:00
------------------------------------------------------------
File: /Dragon.gcode.3mf
Name: Dragon.gcode.3mf
DOS Name: dragon~1.3mf
Size: 2,068,670 bytes
Date: 2024-10-09 23:52:02+00:00
------------------------------------------------------------
File: /parts1.gcode.3mf
Name: parts1.gcode.3mf
DOS Name: parts1~1.3mf
Size: 827,645 bytes
Date: 2024-10-21 23:49:12+00:00
------------------------------------------------------------
File: /Main.gcode.3mf
Name: Main.gcode.3mf
DOS Name: maingc~1.3mf
Size: 0 bytes
Date: 2024-10-21 23:55:02+00:00
------------------------------------------------------------
File: /Wirespool_1.gcode.3mf
Name: Wirespool_1.gcode.3mf
DOS Name: wiresp~1.3mf
Size: 790,004 bytes
Date: 2024-10-21 23:56:28+00:00
------------------------------------------------------------
File: /Wirespool_A.gcode.3mf
Name: Wirespool_A.gcode.3mf
DOS Name: wiresp~2.3mf
Size: 670,850 bytes
Date: 2024-10-21 23:58:06+00:00
------------------------------------------------------------
File: /Wirespool_B.gcode.3mf
Name: Wirespool_B.gcode.3mf
DOS Name: wiresp~3.3mf
Size: 322,783 bytes
Date: 2024-10-21 23:59:14+00:00
------------------------------------------------------------
File: /Wirespool_1_stack.gcode.3mf
Name: Wirespool_1_stack.gcode.3mf
DOS Name: wiresp~4.3mf
Size: 830,635 bytes
Date: 2024-10-23 16:21:42+00:00
------------------------------------------------------------
```

hacked sth together already :) going to drop the code soon when this is proper. Maybe later today.

**Comment by AdrianGarside** [comment]
Does HA have a robust enough UI capabilities to take a list of files, let the user select one and then trigger an action with the choice?

**Reply by Infraviored** [reply]
I would say yes. We just have to get a list of thumbnails of the 3MF files. 
Are these embedded in the 3MF, or is there an additional folder of thumbnails?

It the UI addition becomes boring, I will just create a HTML/JS and embed it in HA.

**Comment by RHolmesLboro** [comment]
they're embedded in the 3mf file, but 3mf is literally just a renamed zip archive and they all have a standard layout so its piss easy to get the image from it.
________________________________

**Comment by j-archit** [comment]
Is this in the works? I've tried using the existing print_project_file action with my A1 to no avail, is it expected to.. not work?

**Reply by AdrianGarside** [reply]
There aren't currently any known bugs with it. But you have to get the provided settings _exactly_ right or the printer will ignore it and/or heat up and then stop. When it comes to the AMS mapping it's close to impossible to get it right because it requires deep knowledge of both what the original ams mappings were for the print and you have to then provide the remapping in an extremely counter-intuitive way. I added new model caching and new card UI in 2.2.0 to make it actually usable, but that's not automatable as it requires you to interact with the card.

**Comment by brunomunizaf** [comment]
Check this out: https://github.com/brunomunizaf/BambuScheduler
